src/util/SimpleTimer.h

```diff
@@ -0,0 +1,192 @@
+// replaces them with a subset of the functionality using Counters internally.
+template <typename Duration> class SimpleTimer
+{
+    static_assert(
+        isDuration<Duration>,
+        "SimpleTimer must be instantiated with a std::chrono::duration");
+    medida::Counter& mSum;
+    medida::Counter& mCount;
+    // Note that we use a counter for `mMax` so it gets displayed in the
+    // metrics, but this is only synced on `syncMax()` to avoid races.
+    medida::Counter& mMaxCounter;
+    std::int64_t mMax;
+
+    std::chrono::steady_clock::time_point mLastUpdate GUARDED_BY(mLock);
+    std::optional<std::chrono::nanoseconds> const mWindowSize GUARDED_BY(mLock);
+
+    // Protects access to mMax and mLastUpdate
+    Mutex mLock;
+
+    void syncMaxUnlocked() REQUIRES(mLock);
+
+  public:
+    // Specify `windowSize` to auto-call `syncMax` on every `Update` that takes
+    // place at least `windowSize` since the last `syncMax`.
+    SimpleTimer(
+        medida::MetricsRegistry& registry, std::string const& domain,
+        std::string const& type, std::string const& baseName,
+        std::optional<std::chrono::nanoseconds> windowSize = std::nullopt);
+
+    SimpleTimer(SimpleTimer<Duration>&& other);
+
+    // Update the value of the max metric to the value of the tracked max and
+    // reset the tracked max
+    void syncMax();
+
+    // Get the value of the internal `count` counter
+    std::int64_t count() const;
+
+    // Record a sample of length duration
+    void Update(std::chrono::nanoseconds duration);
+
+    SimpleTimerContext<Duration> TimeScope();
+};
+
+template <typename Duration> class SimpleTimerContext
+{
+    static_assert(
+        isDuration<Duration>,
+        "SimpleTimerContext must be instantiated with a std::chrono::duration");
+    SimpleTimer<Duration>& mTimer;
+    bool mActive;
+    std::chrono::steady_clock::time_point mStart;
+
+  public:
+    SimpleTimerContext(SimpleTimer<Duration>& timer);
+    ~SimpleTimerContext();
+    void Reset();
+    std::chrono::nanoseconds Stop();
+};
+
+template <typename Duration>
+SimpleTimer<Duration>::SimpleTimer(
+    medida::MetricsRegistry& registry, std::string const& domain,
+    std::string const& type, std::string const& base_name,
+    std::optional<std::chrono::nanoseconds> windowSize)
+    : mSum{registry.NewCounter({domain, type, std::string{base_name} + "sum"})}
+    , mCount{registry.NewCounter(
+          {domain, type, std::string{base_name} + "count"})}
+    , mMaxCounter{registry.NewCounter(
+          {domain, type, std::string{base_name} + "max"})}
+    , mMax{0}
+    , mLastUpdate{std::chrono::steady_clock::now()}
+    , mWindowSize{windowSize}
+{
+}
+
+template <typename Duration>
+SimpleTimer<Duration>::SimpleTimer(SimpleTimer<Duration>&& other)
+    : mSum{other.mSum}
+    , mCount{other.mCount}
+    , mMaxCounter{other.mMaxCounter}
+    , mMax{other.mMax}
+    , mLastUpdate{other.mLastUpdate}
+    , mWindowSize{other.mWindowSize} {};
```

Comment:
I think this move constructor needs to grab `other`'s lock before reaching into it.